Add instructions for changing robocup network. (#265)
NUbots/NUbook: src/book/03-guides/01-main-codebase/04-robocup-setup.mdx

---
section: Guides

> chapter: Main Codebase

Reviewer: It's not quite this. More firmware. Maybe under hardware, since it's most close to Flashing a Robot? Not sure. It could even go in that page.

> title: Robocup Setup

Reviewer: ```suggestion
title: RoboCup Network Setup
```

Author: I was thinking we could have more than just this guide on this page :shrug:.

> description: Setting up the robot for robocup

Reviewer: ```suggestion
description: Setting up the robot networking for robocup
```

> slug: /guides/main/getting-started

Reviewer: ```suggestion
slug: /guides/main/robocup-network
```

authors:

>   - Ysobel Sims (@ysims)
>   - Josephus Paye II (@JosephusPaye)
>   - Mikyla Peters (@miikyla)
>   - Lachlan Court (@LachlanCourt)

Reviewer: ```suggestion
  - Cameron Murtagh (@CMurtagh-LGTM)
```

---

## Networking

1. `cd system/nugusX/etc/systemd/network/` to go to the config folder that system configuration will use for networkd.
2. Rename `40-wlp58s0-robocup.network` to `30-wlp58s0-robocup.network`

> 3. Rename `30-wlp58s0.network` to `40-wlp58s0.network`

Reviewer 2: You could just rename the RoboCup one to be lower than 30 rather than renaming both. Also I think both the old 30 file and the new 30 file will both be around still, and it'll just be lexicographical order of which one runs.

Also it might be worth adding something as to how to undo this after RoboCup, since both of these files will be left behind and you'll need to put it back in "Lab" mode

Reviewer 2: I think the best way might be to swap the two files rather than changing their numbers actually...

Author: Swapping them will be good, the system configuration tool does not delete any old files in these directories so they will stay around on the robot.

4. Double check that network, gateway and dns are correct.

>    Usually the address is `192.168.0.X` or `10.1.0.X` and both gateway and dns end with `3.1`.

Author: `10.1.X.X` and `192.168.X.X` are kinda universal, basically all default subnets are one or the other.

5. `cd ~` to go back to home.

> 5. Add the output of `wpa_passphrase SSID PASSWORD` to `system/default/etc/wpa_supplicant/wpa_supplicant-wlp58s0.conf`

Reviewer: This wouldn't make sense to a lot of people, particularly those running it for the first time. Might as well just say to run 
`wpa_passphrase SSID PASSWORD > system/default/etc/wpa_supplicant/wpa_supplicant-wlp58s0.conf` 
you probably should do an absolute path as well. and say what SSID and PASSWORD meanm and what the whole line does.

Reviewer 2: If wpa_supplicant is in the docker image you could just do something with ./b shell

Author: I don't want people blindly appending to the `wpa_supplicant` config file, I want to make sure they look at it before they add stuff.

>    and add `priority=3` inside the curly brackets on a new line.

Reviewer: probably specify to look for the new entry with the SSID input just before

>    You can use 😎 `vim` or 🤮 `nano` to edit files on the robot.

Reviewer: ```suggestion
   You can use `vim` or `nano` to edit files on the robot.
```

Reviewer 2: boooo :P

---
section: Guides
chapter: Main Codebase
title: Robocup Setup
description: Setting up the robot for robocup
slug: /guides/main/robocup-setup
authors:
    - Cameron Murtagh (@CMurtagh-LGTM)
---

## Networking

> 1. On the robot with a monitor and keyboard, not via `ssh`.

Reviewer 3: Maybe have this as an Alert as it's not really a step but more a method of operation
```suggestion
<Alert type='info'>

Network setup must be completed on the robot with a monitor and keyboard, not via `ssh`.

</Alert>
```

Reviewer 2: If you do this right you can do it by ssh. We always used to. Only had to go to mouse and keyboard when we screwed up :P

2. Rename `system/nugusX/etc/systemd/network/40-wlp58s0-robocup.network`
   to `system/nugusX/etc/systemd/network/30-wlp58s0.network`

> 3. Rename `system/defualt/etc/systemd/30-wlp58s0.network` to `system/default/etc/systemd/40-wlp58s0-robocup.network`

Reviewer 2: ```suggestion
3. Rename `system/default/etc/systemd/30-wlp58s0.network` to `system/default/etc/systemd/40-wlp58s0-robocup.network`
```
You don't actually have to do this step though. You can just rename the robocup one to the 30 one

Also should really do this in the codebase on a RoboCup branch so it persists and gets set every time.
I know you mention this later but I think that should be the normal way of dealing with robocup.
Have the config setup on a RoboCup branch for each robot and just run from there

Author: I think it is clearer if we do *something* with the default one.

4. Double check that `Address`, in the file now called `30-wlp58s0.network`, is in the subnet you are connecting to,
   this is the address of this robot.
   Usually the subnet mask is `255.255.0.0` meaning that the first two numbers define the subnet.
   Usually the address of the subnet is `192.168.X.X` or `10.1.X.X`.
5. Also check that `Gateway` and the first `DNS` are correct in this file, both gateway and dns normally end with

>    `.3.1` and are in the subnet.

Reviewer 2: Normally the RoboCup networks don't have a DNS or gateway

>    `.3.1` and are in the subnet.

Reviewer 4: Isn't `.3.1` highly specific to the network you are connecting to? I feel like saying "normally" here is misleading.

DNS is also irrelevant for the robocup network (there is no internet to connect to and all communication on this network is either broadcast/multicast or to a specific IP).

> 6. Add the output of `wpa_passphrase SSID PASSWORD` to `system/default/etc/wpa_supplicant/wpa_supplicant-wlp58s0.conf`

Reviewer 4: ```sh
wpa_passphrase SSID PASSWORD >> system/default/etc/wpa_supplicant/wpa_supplicant-wlp58s0.conf
```

Or just manually add the entire block to the file
```
network={
       ssid="SSID"
       psk="PASSWORD"
       priority=3
}
```

   where `SSID` is the SSID and `PASSWORD` is the password of the Wi-Fi network you are connecting to.

>    Now add `priority=3` inside the curly brackets on a new line to the entry that you just added.

Reviewer 4: I feel like you should show an example config file here so people know what it should look like

   You can use `vim` or `nano` to edit files on the robot.
7. Run `sudo ./system_configuration`.
8. Reboot the robot `reboot` or restart the services

>    `sudo systemctl restart wpa_supplicant@wlp58s0` and `sudo systemctl restart wpa_supplicant@wlp58s0`.

Reviewer 4: If the network file has also been changed (`/etc/systemd/40-wlp58s0-robocup.network`) you probably also need to do a `sudo systemctl daemon-reload` followed by a `sudo systemctl restart systemd-networkd` before you restart `wpa_supplicant`. Might be a good idea to make `system_configuration` restart the services itself?

description: Setting up the robot for robocup
slug: /guides/main/robocup-setup
authors:
  - Cameron Murtagh (@CMurtagh-LGTM)
---

## Networking

1.  Checkout the branch for robocup and change directory to `modules/tools/SystemConfiguration/data/system`.
2.  Rename `default/etc/systemd/30-wlp58s0.network` to `default/etc/systemd/40-wlp58s0-robocup.network`
3.  Add the output of `wpa_passphrase SSID PASSWORD` to `default/etc/wpa_supplicant/wpa_supplicant-wlp58s0.conf`

>     where `SSID` is the SSID and `PASSWORD` is the password of the Wi-Fi network you are connecting to.

Reviewer 4: ```suggestion
    where `SSID` and `PASSWORD` are the SSID and the password of the Wi-Fi network you are connecting to.
```

    Now add `priority=3` inside the curly brackets on a new line to the entry that you just added.
    The block should look like, where `HASHED_PASSWORD` is a long string of letters and numbers

>     ```
>     network={
>            ssid="SSID"
>            #psk="PASSWORD"
>            psk=HASHED_PASSWORD
>            priority=3
>     }
>     ```

Reviewer 4: I still thing you should just tell people to add this block to the end of `default/etc/wpa_supplicant/wpa_supplicant-wlp58s0.conf`, they have to come open this file in an editor to add the `priority=3`line anyway, and telling them to add the whole block to the end of the file circumvents people accidentally adding the priority line to the wrong block 
```suggestion
    ```
    network={
           ssid="SSID"
           psk="PASSWORD"
           priority=3
    }
    ```
```

Reviewer: Does this work? I remember trying to manually put in the info once before and it wouldn't work until I added it in with `wpa_passphrase` and had the hashed password? But maybe I did something else wrong :shrug:

Reviewer 4: The manpages wpa_supplicant.conf specifically say
```
psk          WPA preshared key used in WPA-PSK mode.  The key is specified as
	     64	hex digits or as an 8-63 character ASCII passphrase.  ASCII
	     passphrases are dynamically converted to a	256-bit	key at runtime
	     using the network SSID, or	they can be statically converted at
	     configuration time	using the wpa_passphrase(8) utility.
```
And this forum post also says this
https://superuser.com/a/679959/1676555

so it should work just fine

Reviewer: this would be better, I'll test and update it

4.  For each robot `X` do steps 5 to 10.
5.  Rename `nugusX/etc/systemd/network/40-wlp58s0-robocup.network`
    to `nugusX/etc/systemd/network/30-wlp58s0.network`.
6.  Double check that `Address`, in the file now called `30-wlp58s0.network`, is in the subnet you are connecting to,
    this is the address of this robot.
    Usually the subnet mask is `255.255.0.0` meaning that the first two numbers define the subnet.
    Usually the address of the robot is `192.168.TEAM_NUMBER.ROBOT_NUMBER` or `10.1.TEAM_NUMBER.ROBOT_NUMBER`.
7.  Normally `Gateway` and `DNS` are not needed for robocup so remove them from the file.
8.  Change directory back to the root of the NUbots repo and install the configs `./b install -co nX`.
    <Alert type='info'>

    When running via `ssh` make sure that you have a monitor and keyboard handy to handle any mistakes.

    </Alert>

9.  Run `sudo ./system_configuration` on the robot.

> 10. Reboot the robot `reboot` or restart the services `sudo systemctl daemon-reload`,
>     `sudo systemctl restart systemd-networkd` and `sudo systemctl restart wpa_supplicant@wlp58s0`.

Reviewer 4: ```suggestion
10. Reboot the robot `reboot` or restart the services `sudo systemctl daemon-reload && sudo systemctl restart systemd-networkd && sudo systemctl restart wpa_supplicant@wlp58s0`.
```

Author: I'm not a fan of just slapping all three commands together, commands linked together with lots of `&&` are quite ugly and, I feel, make it look more complex.

## Networking

1.  Checkout the branch for robocup and change directory to `modules/tools/SystemConfiguration/data/system`.
2.  Rename `default/etc/systemd/30-wlp58s0.network` to `default/etc/systemd/40-wlp58s0-robocup.network`
3.  Add the output of `wpa_passphrase SSID PASSWORD` to `default/etc/wpa_supplicant/wpa_supplicant-wlp58s0.conf`
    where `SSID` is the SSID and `PASSWORD` is the password of the Wi-Fi network you are connecting to.
    Now add `priority=3` inside the curly brackets on a new line to the entry that you just added.
    The block should look like, where `HASHED_PASSWORD` is a long string of letters and numbers
    ```
    network={
           ssid="SSID"
           #psk="PASSWORD"
           psk=HASHED_PASSWORD
           priority=3
    }
    ```
4.  For each robot `X` do steps 5 to 10.
5.  Rename `nugusX/etc/systemd/network/40-wlp58s0-robocup.network`
    to `nugusX/etc/systemd/network/30-wlp58s0.network`.
6.  Double check that `Address`, in the file now called `30-wlp58s0.network`, is in the subnet you are connecting to,
    this is the address of this robot.
    Usually the subnet mask is `255.255.0.0` meaning that the first two numbers define the subnet.

>     Usually the address of the subnet is `192.168.x.x` or `10.1.x.x`.

Reviewer 2: Might be worth saying that the specific IP addresses and subnet masks are provided by RoboCup.
Often we end up with a `192.168.TEAM_NUMBER.ROBOTNUMBER` but different years the format has varied
